controllers/core/openstackversion_controller.go

```diff
@@ -0,0 +1,288 @@
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+
+		err := versionHelper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+
+	instance.Status.Conditions.Set(condition.FalseCondition(
+		corev1beta1.OpenStackVersionInitialized,
+		condition.RequestedReason,
+		condition.SeverityInfo,
+		corev1beta1.OpenStackVersionInitializedReadyRunningMessage))
+
+	instance.Status.AvailableVersion = &envAvailableVersion
+	defaults := openstack.InitializeOpenStackVersionImageDefaults(ctx, envContainerImages)
+	// store the defaults for the currently available version
+	if instance.Status.ContainerImageVersionDefaults == nil {
+		instance.Status.ContainerImageVersionDefaults = make(map[string]*corev1beta1.ContainerDefaults)
+	}
+	instance.Status.ContainerImageVersionDefaults[envAvailableVersion] = defaults
+
+	// calculate the container images for the target version
+	if val, ok := instance.Status.ContainerImageVersionDefaults[instance.Spec.TargetVersion]; !ok {
+		Log.Info("Target version not found in defaults", "targetVersion", instance.Spec.TargetVersion)
+		return ctrl.Result{}, nil
+	} else {
+		instance.Status.ContainerImages = openstack.GetContainerImages(ctx, val, *instance)
+	}
+
+	instance.Status.Conditions.MarkTrue(
+		corev1beta1.OpenStackVersionInitialized,
+		corev1beta1.OpenStackVersionInitializedReadyMessage)
+
+	// greenfield deployment //FIXME check dataplane here too
+	if controlPlane == nil || controlPlane.Status.DeployedVersion == nil {
+		Log.Info("No controlplane or controlplane is not deployed")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO minor update for OVN Dataplane in progress
+
+	// minor update for OVN Controlplane in progress
+	if instance.Status.DeployedVersion != nil && instance.Spec.TargetVersion != *instance.Status.DeployedVersion {
+		if controlPlane.Status.Conditions.IsFalse(condition.ReadyCondition) {
+			Log.Info("Minor update Controlplane is not in ready condition, waiting")
+			return ctrl.Result{}, nil
+		}
+		if !compareStringPointers(controlPlane.Status.ContainerImages.OvnControllerImage, instance.Status.ContainerImages.OvnControllerImage) ||
+			controlPlane.Status.Conditions.IsFalse(corev1beta1.OpenStackVersionMinorUpdateOVNControlplane) {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				corev1beta1.OpenStackVersionMinorUpdateOVNControlplane,
+				condition.RequestedReason,
+				condition.SeverityInfo,
+				corev1beta1.OpenStackVersionMinorUpdateReadyRunningMessage))
+			Log.Info("Minor update for OVN Controlplane in progress")
+			return ctrl.Result{}, nil
+		}
+		instance.Status.Conditions.MarkTrue(
+			corev1beta1.OpenStackVersionMinorUpdateOVNControlplane,
+			corev1beta1.OpenStackVersionMinorUpdateReadyMessage)
+
+		// minor update for Controlplane in progress
+		// we only check keystone here as it will only get updated during this phase
```

Comment:
Why don't we check all the service container images and the Ready condition of the OpenStackControlPlane instead to wait for all the services to be redeployed with the new image?

Reply:
Or do we need only keystone to be updated before we do the dataplane update?

Reply:
the isReady conditions need to be carefully reviewed before we can consolidate this. I would like to check them all, and that would certainly be more complete. For now this was just sort of a spot check.